pose-detection/demos/live_video/src/renderer_webgpu.js

```diff
@@ -0,0 +1,369 @@
+  }
+
+  draw(video, tensor, canvasInfo) {
+    this.canvasInfo = canvasInfo;
+    const videoCommanderBuffer = this.drawTexture(video);
+    const poseCommanderBuffer = this.drawPose(tensor);
+    this.device.queue.submit([videoCommanderBuffer, poseCommanderBuffer]);
+  }
+
+  getPoseShader(tensorShapeX) {
+    const vertexShaderCode = `
+struct Uniforms {
+  offsetX : f32,
+  offsetY : f32,
+  scaleX : f32,
+  scaleY : f32,
+  width : f32,
+  height : f32,
+}
+
+struct Pose {
+  keypoints : array<vec2<f32>, ${tensorShapeX}>
+}
+
+@binding(0) @group(0) var<uniform> uniforms : Uniforms;
+@binding(1) @group(0) var<storage> pose : Pose;
+@vertex
+fn main(
+  @builtin(vertex_index) VertexIndex : u32
+) -> @builtin(position) vec4<f32> {
+  let rawY = (pose.keypoints[VertexIndex].x + uniforms.offsetY) * uniforms.scaleY / uniforms.height;
+  let rawX  = (pose.keypoints[VertexIndex].y + uniforms.offsetX) * uniforms.scaleX / uniforms.width;
+  var x = rawX * 2.0 - 1.0;
+  var y = 1.0 - rawY * 2.0;
+  return vec4<f32>(x, y, 1.0, 1.0);
+}
+    `;
+    const fragmentShaderCode = `
+@fragment
+fn main() -> @location(0) vec4<f32> {
+  return vec4<f32>(1.0, 0.0, 0.0, 1.0);
+}
+    `;
+    return [vertexShaderCode, fragmentShaderCode];
+  }
+
+  initDrawPose(tensor) {
+    // Only 2d tensor whose last dimension is 2 is supported.
+    if (tensor == null || tensor.shape.length !== 2 || tensor.shape[1] !== 2) {
+      throw new Error('Tensor is null or tensor shape is not supported!');
+    }
+
+    // pose-detection supports 17 body parts:
+    // 'nose', 'left_eye', 'right_eye', 'left_ear', 'right_ear',
+    // 'left_shoulder', 'right_shoulder', 'left_elbow', 'right_elbow',
+    // 'left_wrist', 'right_wrist', 'left_hip', 'right_hip', 'left_knee',
+    // 'right_knee', 'left_ankle', 'right_ankle'. This demo draws the first
+    // five.
+    const poseIndexArray = new Uint32Array([3, 1, 1, 0, 0, 2, 2, 4]);
+    this.poseIndexCount = poseIndexArray.length;
+
+    if (this.indexBuffer == null) {
+      this.indexBuffer = this.createBuffer(
+          GPUBufferUsage.INDEX, poseIndexArray.byteLength, poseIndexArray);
+    }
+    if (this.uniformBuffer == null) {
+      this.uniformBuffer = this.createBuffer(
+          GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
+          this.canvasInfo.length * 4);
+    }
+    if (this.posePipeline == null) {
+      this.posePipeline = this.createPosePipeline(tensor.shape[0]);
+    }
+    const bindings = [
+      {
+        buffer: this.uniformBuffer,
+        offset: 0,
+        size: this.canvasInfo.length * 4,
+      },
+      {
+        buffer: tensor.dataToGPU().buffer,
+        offset: 0,
+        size: byteSizeFromShape(tensor.shape),
+      }
+    ];
+    return this.device.createBindGroup({
+      layout: this.posePipeline.getBindGroupLayout(0),
+      entries: bindings.map((b, i) => ({binding: i, resource: b})),
+    });
+  }
+
+  drawPose(tensor) {
+    const poseBindGroup = this.initDrawPose(tensor);
```

Comment:
Can the bindGroup also be cached so that we don't need to call `createBindGroup` every time?
And maybe we can move the `initxxx` to setup stage.

Reply:
The GPUBuffer from tensor is different after each predict.

Reply:
The GPUBuffer from tensor is different after each predict.